### PR TITLE
KAFKA-15174: Ensure CommitAsync propagate the exception to the user

### PR DESCRIPTION
**Pull request by philipnee**
The PR covers a few important points:
1. Exception handling: We should be thrown RetriableCommitException when the commit exception is retriable.  We should throw FencedIdException on commit and poll similar to the current implementation.  Other errors should be thrown as it is.
2. Callback invocation: The callbacks need to be invoked on the main/application thread; however, the future is completed in the background thread.  To achieve this, I created an Invoker class with a queue, so that these callback can be invoked during the consumer.poll()

Note: One change I made is to remove the DefaultOffsetCommit callback.  Since the callback is purely for logging, I think it is reasonable for us to move the logging to the background thread instead of relying on the application thread to trigger the logging.

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java [hunk 1]
```diff
@@ -384,12 +384,10 @@ public void commitAsync(Map<TopicPartition, OffsetAndMetadata> offsets, OffsetCo
         final OffsetCommitCallback commitCallback = callback == null ? new DefaultOffsetCommitCallback() : callback;
         future.whenComplete((r, t) -> {
             if (t != null) {
-                commitCallback.onComplete(offsets, new KafkaException(t));
+                commitCallback.onComplete(offsets, (Exception) t);
             } else {
                 commitCallback.onComplete(offsets, null);
             }
-        }).exceptionally(e -> {
-            throw new KafkaException(e);
```
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java]: I had a question about this code (even from before this change)—On which thread will the user's callback be executed?

I'm pretty sure that the event's `Future` is completed on the background thread, right? If so, that means that the callback passed to `whenComplete` is going to be completed on the background thread. And since it then calls the user's callback, it will also be called on the background thread.

Thoughts?

Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java]: thanks for catching this. I think you are right.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java [hunk 2]
```diff
@@ -154,6 +158,8 @@ public class PrototypeAsyncConsumer<K, V> implements Consumer<K, V> {
     private AtomicBoolean fencedInstance = new AtomicBoolean(false);
     private final Optional<String> groupInstanceId;
 
+    private ExecutorService callbackExecutor = Executors.newSingleThreadExecutor();
```
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java]: This is a bit weird, no? In the current model, I suppose that the callbacks are executed by the application thread when poll() is called. Should we do the same here?

**Comment by dajac** [comment]
Does commitSync have the same issue?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java (hunk 2 above)
Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java]: Sorry - this is wrong.  The way I wanted to do it in this PR is to register the callbacks, when the future is completed, to the application thread.  And we invoke them when we call poll.  I believe the callback invocation is currently invoked on consumer.poll(), because this is where network clients are polled.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java [hunk 3]
```diff
@@ -1096,4 +1114,47 @@ private void subscribeInternal(Collection<String> topics, Optional<ConsumerRebal
...
+    }
+
+    /**
+     * Utility class that helps the application thread to invoke user registered callbacks such as
+     * {@link OffsetCommitCallback}.  This is achieved by having the background thread to register a runnable to the
+     * invoker upon the future completion, and execute the callbacks when user polls the consumer.
+     */
+    private static class CallbackInvoker {
+        // Thread-safe queue to store callbacks
+        private final BlockingQueue<Runnable> callbackQueue = new LinkedBlockingQueue<>();
```
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java]: I also wonder if we really need this new queue. Could we rely on the existing queue between the background and the foreground threads? I haven't look into it myself but I wonder...

Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java]: Possible.  However, the OffsetCommitCallback was not registered to the background thread, that why I added an invoker.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java [hunk 4]
```diff
@@ -381,20 +394,33 @@ public void commitAsync(OffsetCommitCallback callback) {
...
+                    fencedInstance.set(true);
+                    callback.onComplete(offsets, (Exception) t);
+                } else if (t != null) {
+                    callback.onComplete(offsets, (Exception) t);
+                } else {
+                    callback.onComplete(offsets, null);
+                }
+            };
+            invoker.submit(task);  // Store the callback task for future execution
         });
```
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java]: The code passed to `whenComplete` will be run on the background thread. Therefore, it's "acceptable" (IMO) to enqueue a new event onto the background event queue. Then, when it's later run in the background event processor, the event processing should trigger the code you have here.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java [hunk 5]
```diff
@@ -326,6 +337,8 @@ public PrototypeAsyncConsumer(LogContext logContext,
      */
     @Override
     public ConsumerRecords<K, V> poll(final Duration timeout) {
+        maybeInvokeCallbacks();
+        maybeThrowFencedInstanceException();
```
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java]: Ideally, these checks should be added to the logic in `BackgroundEventProcessor`. Right now, `BackgroundEventProcessor` is a separate class, which makes it a bit awkward, since you need access to the consumer's private instance variables. In a PR that I have in review, the `BackgroundEventProcessor` is an inner class of the consumer, making getting and setting the instance variable values trivial.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumerTest.java [hunk 6]
```diff
@@ -168,6 +200,45 @@ public void testCommitted_ExceptionThrown() {
...
+    }
+
+    private static class MockCommitCallback implements OffsetCommitCallback {
+        public int invoked = 0;
+        public Exception exception = null;
+        public String completionThread;
+
+        @Override
+        public void onComplete(Map<TopicPartition, OffsetAndMetadata> offsets, Exception exception) {
+            System.out.println("invoke");
```
Comment (kirktrue) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumerTest.java]: We should remove the `System.out.println()` from committed code.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java (hunk 3 above)
Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java]: In fact - using the backgroundEventQueue would not be feasible. I just realized the callbacks need to be invoked on all commit calls, and poll.  I don't know if there's a better way to do it.

**Comment by philipnee** [comment]
Hi @dajac @kirktrue - Thanks for offering valuable feedbacks to this PR, much appreciated. I'm not sure if there is a way to avoid using a separated collection on the main thread to store the callback, because we need to register these callbacks somewhere to ensure they are invoked on the main thread during some of the API calls. 

Previously, David suggested that we could try to use the background queue; however, this is not really a feasible solution because we don't want the consumer to poll the queue during a commit (which could spit out a bunch of errors and rebalance events). Any suggestions?

I would like to propose a KIP in the future, to enforce the callback invocation to only happen on poll. But this is out of the scope of this PR.

File: core/src/test/scala/integration/kafka/api/BaseAsyncConsumerTest.scala [hunk 7]
```diff
@@ -16,7 +16,7 @@
  */
 package kafka.api
 
-import kafka.utils.TestUtils.waitUntilTrue
+import kafka.utils.TestUtils.{waitUntilTrue}
```
Comment (lucasbru) [inline core/src/test/scala/integration/kafka/api/BaseAsyncConsumerTest.scala]: Do we need curly brackets?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java [hunk 8]
```diff
@@ -147,6 +152,10 @@ public class PrototypeAsyncConsumer<K, V> implements Consumer<K, V> {
     // to keep from repeatedly scanning subscriptions in poll(), cache the result during metadata updates
     private boolean cachedSubscriptionHasAllFetchPositions;
     private final WakeupTrigger wakeupTrigger = new WakeupTrigger();
+    private final AtomicBoolean fencedInstance = new AtomicBoolean(false);
```
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java]:
```suggestion
    private volatile boolean isFenced = false;
```

Any reason this needs to be anything other than a `boolean`?

**Review by kirktrue — CHANGES_REQUESTED** [review]
Thanks for the updates, @philipnee. I can appreciate the game of Twister you're having to play to get this to work.

Take a look at the forthcoming [changes to `PrototypeAsyncConsumer` in my `groupMetadata` pull request](https://github.com/kirktrue/kafka/blob/KAFKA-15281-implement-groupmetadata/clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java#L1115). With those changes, I would imagine that you could add a new `BackgroundEvent` and enqueue it in the background event queue in your callback in `commitAsync`:

```java
public void commitAsync(Map<TopicPartition, OffsetAndMetadata> offsets, OffsetCommitCallback callback) {
    CompletableFuture<Void> future = commit(offsets, false);
    future.whenComplete((r, t) -> {
        // Note: this Future is being completed by the background thread, so only enqueue an event
        // for the application thread
        backgroundEventHandler.add(new OffsetCommitBackgroundEvent(callback, offsets, t));
    });
}
```

Then you'd update the inner `BackgroundEventProcessor` to add a new `process` method like this:

```java
private void process(final OffsetCommitBackgroundEvent event) {
    Exception exception = event.exception();
    OffsetCommitCallback callback = event.callback();
    Map<TopicPartition, OffsetAndMetadata> offsets = event.offsets();

    if (exception instanceof RetriableException) {
        if (callback != null)
            callback.onComplete(offsets, new RetriableCommitFailedException(exception));
        else
            log.error("Offset commit with offsets {} failed", offsets, t);

        return;
    }

    if (exception instanceof FencedInstanceIdException)
        fencedInstance = true;

    if (callback != null)
        callback.onComplete(offsets, exception);
    else
        log.error("Offset commit with offsets {} failed", offsets, t);
}
```

The `BackgroundEventProcessor` is already invoked inside `poll()` and if the event is added to the queue and the `process` method is added similar to the above, that should do the trick.

**Comment by philipnee** [comment]
Hi @kirktrue Thanks for taking time reviewing the PR: Just to quickly respond to your comment.

The BackgroundEventProcessor is already invoked inside poll() and if the event is added to the queue and the process method is added similar to the above, that should do the trick.

The callback also needs to be invoked on all commit calls, if we decide to stick with the current behavior, which I assume we do.  (Even though I personally would like to conslidate all invocation on poll only, to make things consistent).

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumerTest.java (hunk 6 above)
Comment (philipnee) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumerTest.java]: Java debugging 101: use println

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java [hunk 9]
```diff
@@ -147,6 +151,10 @@ public class PrototypeAsyncConsumer<K, V> implements Consumer<K, V> {
     // to keep from repeatedly scanning subscriptions in poll(), cache the result during metadata updates
     private boolean cachedSubscriptionHasAllFetchPositions;
     private final WakeupTrigger wakeupTrigger = new WakeupTrigger();
+    private volatile boolean isFenced = false;
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java]: I'm wondering why we need this member. It looks like the volatile boolean is written and read only from the application thread. Could it then just be a boolean? But if so, could we just throw the exception directly and avoid the indirection?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java [hunk 10]
```diff
@@ -278,7 +286,8 @@ public PrototypeAsyncConsumer(final Time time,
         }
     }
 
-    public PrototypeAsyncConsumer(LogContext logContext,
+    // Visible for testing
+    PrototypeAsyncConsumer(LogContext logContext,
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java]: You will have to fix the indentation of the other parameters.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java [hunk 11]
```diff
@@ -381,21 +391,28 @@ public void commitAsync(OffsetCommitCallback callback) {
...
             }
-        }).exceptionally(e -> {
-            throw new KafkaException(e);
+
+            invoker.submit(new OffsetCommitCallbackTask(callback, offsets, (Exception) t));
         });
     }
 
     // Visible for testing
     CompletableFuture<Void> commit(Map<TopicPartition, OffsetAndMetadata> offsets, final boolean isWakeupable) {
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java]: nit: some parameters final, some non-final

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java [hunk 12]
```diff
@@ -381,21 +391,28 @@ public void commitAsync(OffsetCommitCallback callback) {
...
+        maybeThrowFencedInstanceException();
+        maybeInvokeCallbacks();
         maybeThrowInvalidGroupIdException();
+
+        if (offsets.isEmpty()) {
+            return CompletableFuture.completedFuture(null);
+        }
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java]: Maybe not related to this concrete PR, but the original commit code has some logic around updating the last seen leader epochs here, can we omit it, or is it a follow-up task?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java [hunk 13]
```diff
@@ -1010,15 +1031,9 @@ private void updateLastSeenEpochIfNewer(TopicPartition topicPartition, OffsetAnd
...
-    private class DefaultOffsetCommitCallback implements OffsetCommitCallback {
-        @Override
-        public void onComplete(Map<TopicPartition, OffsetAndMetadata> offsets, Exception exception) {
-            if (exception != null)
-                log.error("Offset commit with offsets {} failed", offsets, exception);
-        }
-    }
-
     boolean updateAssignmentMetadataIfNeeded(Timer timer) {
+        maybeInvokeCallbacks();
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java]: Sometimes we check for the fencedInstance first, sometimes we invoke the callbacks first.

I think in a previous version of the PR and in the original consumer we invoke at the beginning of `poll`. Why did you move this into `updateAssignmentMetadataIfNeeded`?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java [hunk 14]
```diff
@@ -1096,4 +1111,76 @@ private void subscribeInternal(Collection<String> topics, Optional<ConsumerRebal
...
\ No newline at end of file
+    private void maybeThrowFencedInstanceException() {
+        if (isFenced) {
+            throw new FencedInstanceIdException("Get fenced exception for group.instance.id " +
+                groupInstanceId.orElse("null"));
+        }
+    }
+
+    // Visible for testing
+    void maybeInvokeCallbacks() {
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java]: nit: how about `maybeInvokeCommitCallbacks`?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java [hunk 15]
```diff
@@ -1096,4 +1111,76 @@ private void subscribeInternal(Collection<String> topics, Optional<ConsumerRebal
...
+    }
+
+    // Visible for testing
+    int callbacks() {
+        return invoker.callbackQueue.size();
+    }
+
+    /**
+     * Utility class that helps the application thread to invoke user registered {@link OffsetCommitCallback}. This is
+     * achieved by having the background thread to register a {@link OffsetCommitCallbackTask} to the invoker upon the
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java]:
```suggestion
     * achieved by having the background thread register a {@link OffsetCommitCallbackTask} to the invoker upon the
```

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java [hunk 16]
```diff
@@ -1096,4 +1111,76 @@ private void subscribeInternal(Collection<String> topics, Optional<ConsumerRebal
...
+        public void submit(final OffsetCommitCallbackTask callback) {
+            try {
+                callbackQueue.offer(callback);
+            } catch (Exception e) {
+                log.error("Failed to enqueue OffsetCommitCallback", e);
+            }
+        }
+
+        public void executeCallbacks() {
+            LinkedList<OffsetCommitCallbackTask> callbacks = new LinkedList<>();
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java]: Why do we need to create a copy of the queue here? Are we concerned about consumer invocations inside the callback? Then maybe a put a comment here.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java [hunk 17]
```diff
@@ -1096,4 +1111,76 @@ private void subscribeInternal(Collection<String> topics, Optional<ConsumerRebal
...
+            while (!callbacks.isEmpty()) {
+                OffsetCommitCallbackTask callback = callbacks.poll();
+                if (callback != null) {
+                    callback.invoke();
+                }
+            }
+        }
+    }
+
+    private class OffsetCommitCallbackTask {
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java]: Should be a static class.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java [hunk 18]
```diff
@@ -1096,4 +1111,76 @@ private void subscribeInternal(Collection<String> topics, Optional<ConsumerRebal
...
+    int callbacks() {
+        return invoker.callbackQueue.size();
+    }
+
+    /**
+     * Utility class that helps the application thread to invoke user registered {@link OffsetCommitCallback}. This is
+     * achieved by having the background thread to register a {@link OffsetCommitCallbackTask} to the invoker upon the
+     * future completion, and execute the callbacks when user polls/commits/closes the consumer.
+     */
+    private class OffsetCommitCallbackInvoker {
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java]: Should be a static class.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java [hunk 19]
```diff
@@ -1096,4 +1111,76 @@ private void subscribeInternal(Collection<String> topics, Optional<ConsumerRebal
...
+        // Thread-safe queue to store callbacks
+        private final BlockingQueue<OffsetCommitCallbackTask> callbackQueue = new LinkedBlockingQueue<>();
+
+        public void submit(final OffsetCommitCallbackTask callback) {
+            try {
+                callbackQueue.offer(callback);
+            } catch (Exception e) {
+                log.error("Failed to enqueue OffsetCommitCallback", e);
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java]: Why do we expect an exception inside offer?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java [hunk 20]
```diff
@@ -712,6 +729,10 @@ private void close(Duration timeout, boolean swallowException) {
         if (applicationEventHandler != null)
             closeQuietly(() -> applicationEventHandler.close(timeout), "Failed to close application event handler with a timeout(ms)=" + timeout, firstException);
 
+        // Invoke all callbacks after the background thread exists in case if there are unsent async
+        // commits
+        maybeInvokeCallbacks();
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java]: Why do we not check for fenced exceptions here?

The original consumer attempts to complete all in-flight async commit requests within `timeout` ms. Will closing the background thread take care of this?

**Review by lucasbru — COMMENTED** [review]
Thanks for the PR! I left some questions / comments.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java (hunk 13 above)
Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java]: Hi, to your first question, it was an oversight, so i fixed it. to your second question: I moved it down for a couple of reasons, one is we want the callback invocation to be included in the poll timer, and two, i wanted to consolidate the logic with the backgroundEventProcess.poll to make its behavior similar to the current consumer.  The current consumer poll `ConsumerCoordinator` class in the `updateAssignmentMetadataIfNeeded`.  

the `ConsumerCoordinator#poll` actually performs the callback invocation first, which, first checks the fencedInstaceId. See the code below:

```
public boolean poll(Timer timer, boolean waitForJoinGroup) {
        maybeUpdateSubscriptionMetadata();

        invokeCompletedOffsetCommitCallbacks();
        ...
}
```

and 
```
void invokeCompletedOffsetCommitCallbacks() {
        if (asyncCommitFenced.get()) {
            throw new FencedInstanceIdException("Get fenced exception for group.instance.id "
                + rebalanceConfig.groupInstanceId.orElse("unset_instance_id")
                + ", current member.id is " + memberId());
        }
        while (true) {
            OffsetCommitCompletion completion = completedOffsetCommits.poll();
            if (completion == null) {
                break;
            }
            completion.invoke();
        }
    }
```

Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java]: The `backgroundEventProcessor.process` is very similar to `ConsumerCoordinator#poll`; however, because of the implementation difference, the two implementations will differ slightly.  The change here is to ensure the two behave as similar as possible.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java (hunk 14 above)
Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java]: great!

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java (hunk 16 above)
Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java]: Thanks, I checked the current implementation, we don't need to do this.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java (hunk 9 above)
Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java]: Thanks, we don't need this to be volatile.  In terms of exception throwing, I think once the instance is fenced, the we want to keep throwing if the user tries to commit or poll, because the expectation is that the user should close and reconfigure the instance.  See the original implementation

```
void invokeCompletedOffsetCommitCallbacks() {
        if (asyncCommitFenced.get()) {
            throw new FencedInstanceIdException("Get fenced exception for group.instance.id "
                + rebalanceConfig.groupInstanceId.orElse("unset_instance_id")
                + ", current member.id is " + memberId());
        }
         ...
}
```

Here asyncCommitFenced is set during the callback failure in the original implementation.
```
if (commitException instanceof FencedInstanceIdException) {
                    asyncCommitFenced.set(true);
}
```

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java (hunk 12 above)
Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java]: Thanks, I'll add a PR for investigation. I think metadata was not in the implementation at the time the function was implemented.

Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java]: Here is the Jira: https://issues.apache.org/jira/browse/KAFKA-15803

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java (hunk 17 above)
Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java]: Hi, it is not static for the reason of the use of isFenced.  Let me know what do you think.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java (hunk 18 above)
Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java]: Similar to OffsetCommitCallbackTask

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java (hunk 20 above)
Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java]: When the consumer is closing, we don't really need to throw a fenced id exception, because when the consumer is fenced we want the consumer to close right?

To your second question: yes, the work is inflight, KAFKA-15327.

**Comment by philipnee** [comment]
Hi @lucasbru - Thanks for your feedback, the PR was updated.  Would you have time to go over the comments again? Thanks!

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java (hunk 20 above)
Comment (lucasbru) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java]: The fenced exception would indicate here that the last commit did not complete, correct? I think you are right, throwing the exception is not strictly required for correctness here; at the same time, when investigating what was going on the client during a fencing situation, it would be valuable information that the commit failed due to fencing, even if we are shutting down. If we can tell that from non-DEBUG logs, I'd say it's fine though, I haven't checked.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java (hunk 9 above)
Comment (lucasbru) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java]: Make sense, thanks!

**Comment by lucasbru** [comment]
Hi @philipnee. Looks good, I think all my comments got addressed. I see a comment from @kirktrue about using the background event queue instead of a separate invoker queue, did we resolve this conversation? It seems to make a lot of sense to me, if at all possible.

**Comment by philipnee** [comment]
Hi @lucasbru - Yes. It is not quite feasible to use the backgroundEventQueue because of the behavior of the current API, i.e., the callback needs to be invoked ouside of the poll call.  If we use the backgroundEventQueue, then the operation can become quite complicated, i.e. we will need to scan through the queue, execute the callback, and remove the callback from the queue.  @kirktrue - mind comment here?

**Comment by kirktrue** [comment]
@lucasbru As @philipnee said, the two mechanisms appear similar and do have overlap. However, not only is it "awkward" to mix them, but it raises correctness questions.

`poll()` processes  _all_ of the events in the queue from the background thread. However, when the user calls `commit()`, we want to execute _only_ the commit callbacks. As a result, we would have to _filter and remove_ the relevant events from the background thread.

The questions that arose:

* What if, when processing the commit callbacks in `commit()`, we notice that there are error events in the queue? Do we skip them? Execute them? Or something else?
* Will removing the events from the background event queue in `commit()` cause unforeseen issues later when processing the events in `poll()`?

The answer may be that it's totally fine. We kind of "punted" by doing it this way, not only because of those questions, but because it seemed a bit awkward 🤷‍♂️

**Comment by lucasbru** [comment]
Okay. Sounds like using the background events queue wouldn't strictly have downsides, but also not have so many upsides. I agree that picking things from a queue sounds awkward, so let's stick with a separate queue then.

@philipnee we have to fix a spotbugs problem before we can merge this

**Review by lucasbru — APPROVED** [review]

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java (hunk 9 above)
Comment (lucasbru) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java]: @philipnee Can we remove the volatile if we do not need this?

**Comment by philipnee** [comment]
Hi @lucasbru - Addressed your comment about the volatile.  I also modify the invoker.submit to catch some java exception.  Please review.  Terrible sorry but I needed to rebase to the current trunk so here are the two new commits: [8c133b8](https://github.com/apache/kafka/pull/14680/commits/8c133b87d6c8cd6b87d009c91d6955e2d23c01e5) and [78b3cfa](https://github.com/apache/kafka/pull/14680/commits/78b3cfa6710f9287b3ff70f50ce911052b4ff313)

**Comment by lucasbru** [comment]
@philipnee there is a suspicious test failure for `org.apache.kafka.clients.consumer.KafkaConsumerTest`, could you have a look?

**Comment by philipnee** [comment]
hm - why is it sending heartbeat? looking

**Comment by philipnee** [comment]
I rebased the trunk and was unable to reproduce the failure locally.  I wonder if the tests is flaky and was caused by the delegate PR.

**Comment by philipnee** [comment]
I think the test failures was probably caused by heartbeat bubbled up to the consumer subsequently throwing an exception

**Comment by philipnee** [comment]
@lucasbru - I think we probably need to refactor KafkaConsumerTest again because most of the tests don't really apply to the async KafkaConsumer due to different protocol use etc.  I disabled a couple more tests in the last commit because they were trying to trigger rebalance, and things like

```
        prepareRebalance(client, node, assignor, singletonList(tp0), null);
```
don't necessary apply to the new consumer anymore.

**Comment by philipnee** [comment]
The most recent build passed - let me re-trigger the tests to make sure.

**Comment by philipnee** [comment]
hi @lucasbru - I think the KafkaConsumerTest failures were caused by some test we forgot to disabled.  The last 2 runs seem to be ok (well, still flaky but not failing on that specific one).  I retriggerred it again.  But I think we can merge it.

**Comment by lucasbru** [comment]
Scala compilation in 2.12 is broken on master, which is fixed in https://github.com/apache/kafka/pull/14786, otherwise this is looking good

**Comment by philipnee** [comment]
Thanks @lucasbru

**Comment by lucasbru** [comment]
Merged trunk. Test failures are unrelated
